pkg/calculator/experimentcalc/experiment_calculator.go

```diff
@@ -0,0 +1,650 @@
+	ctx context.Context,
+	req *eventcounter.GetExperimentEvaluationCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	evaluationCount, err := e.eventCounterClient.GetExperimentEvaluationCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range evaluationCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) getGoalCount(
+	ctx context.Context,
+	req *eventcounter.GetExperimentGoalCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	goalCount, err := e.eventCounterClient.GetExperimentGoalCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range goalCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) calcGoalResult(
+	ctx context.Context,
+	evalVariationCounts,
+	goalVariationCounts map[string]*eventcounter.VariationCount,
+	baseVid string,
+) *eventcounter.GoalResult {
+	goalResult := &eventcounter.GoalResult{}
+	length := len(goalVariationCounts)
+	vids := make([]string, 0, length)
+	goalUc, evalUc := make([]int64, 0, length), make([]int64, 0, length)
+	vrs := make(map[string]*eventcounter.VariationResult, length)
+	valueMeans, valueVars := make([]float64, 0, length), make([]float64, 0, length)
+	baselineIdx, loopIdx := 0, 0
+
+	for vid, goalVariationCount := range goalVariationCounts {
+		if _, ok := evalVariationCounts[vid]; !ok {
+			return goalResult
+		}
+		vids = append(vids, vid)
+		evalVariationCount := evalVariationCounts[vid]
+		vr := &eventcounter.VariationResult{
+			VariationId:     vid,
+			ExperimentCount: copyVariationCount(goalVariationCount),
+			EvaluationCount: copyVariationCount(evalVariationCount),
+		}
+		evalUc = append(evalUc, evalVariationCount.UserCount)
+		goalUc = append(goalUc, goalVariationCount.UserCount)
+		valueMeans = append(valueMeans, goalVariationCount.ValueSumPerUserMean)
+		valueVars = append(valueVars, goalVariationCount.ValueSumPerUserVariance)
+
+		goalResult.VariationResults = append(goalResult.VariationResults, vr)
+		vrs[vid] = vr
+		if vid == baseVid {
+			baselineIdx = loopIdx
+		}
+		loopIdx++
+	}
+	// Skip the calculation if evaluation count is less than goal count.
+	for i := 0; i < len(evalUc); i++ {
+		if evalUc[i] < goalUc[i] {
+			return goalResult
+		}
+	}
+
+	cvrResult, sampleErr := e.binomialModelSample(ctx, vids, goalUc, evalUc, baselineIdx)
+	if sampleErr != nil {
+		calculationCounter.WithLabelValues(calculateFail).Inc()
+		e.logger.Error("BinomialModelSample error",
+			log.FieldsFromImcomingContext(ctx).AddFields(
+				zap.Error(sampleErr),
+				zap.String("baseVariationID", baseVid),
+			)...,
+		)
+		return goalResult
```

Comment:
Same here.
https://github.com/bucketeer-io/bucketeer/pull/395#discussion_r1236424262

Reply:
The original Python code logic [here](https://github.com/bucketeer-io/bucketeer/blob/main/python/src/lib/calculator/job/calculate_experiments.py#L472)

Same as  https://github.com/bucketeer-io/bucketeer/pull/395#discussion_r1236570075